### DISCUSSION TITLE
Q: Why do I get a "FlexColorScheme WARNING: primaryLightRef is null, but primary is defined"?

### DISCUSSION THREAD

**Discussion by maxfrees** [rydmike/flex_color_scheme#269] [Q&A] [answered]
I/flutter (25901): FlexColorScheme WARNING: primaryLightRef is null, but primary is defined, primaryLightRef may be needed. Set primaryLightRef to the same color as primary in your LIGHT theme mode setup to get correct values for "fixed" colors for your dark ColorScheme or to seed with same color as your light theme primary. If primary is already set to same color as light mode primary and it is only used as seed in dark mode, defining primaryLightRef is not necessary.

![image](https://github.com/user-attachments/assets/d4fe63c1-f147-4828-b380-a48d109ef54d)

**Comment by rydmike** [comment, marked as answer]
Hi @maxfrees, thanks for your question.

Yes, this is an expected informational debug print, it is only shown/printed in debug mode build and only when there might be a point in doing so. It is intended to make you aware of potentially incomplete configuration to make a matching dark mode ColorScheme that follows the Material-3 design intent.


This is what triggers it:

```dart
    // In debug mode warn to console about missing LightRef colors when they
    // are used in a combo that may require them.
    if (kDebugMode) {
      if (primary != null &&
          primaryLightRef == null &&
          ((!seed.useKeyColors &&
                  (fixedColorStyle ?? FlexFixedColorStyle.computed) ==
                      FlexFixedColorStyle.computed) ||
              seed.useKeyColors)) {
        debugPrint('FlexColorScheme WARNING: primaryLightRef is null, but '
            'primary is defined, primaryLightRef may be needed. '
            'Set primaryLightRef to the same color '
            'as primary in your LIGHT theme mode setup to get correct values '
            'for "fixed" colors for your dark ColorScheme or to seed with '
            'same color as your light theme primary. If primary is already '
            'set to same color as light mode primary and it is only used as '
            'seed in dark mode, defining primaryLightRef is not necessary.');
      }
      if (secondary != null &&
          secondaryLightRef == null &&
          ((!seed.useKeyColors &&
                  (fixedColorStyle ?? FlexFixedColorStyle.computed) ==
                      FlexFixedColorStyle.computed) ||
              seed.useKeyColors)) {
        debugPrint('FlexColorScheme WARNING: secondaryLightRef is null, but '
            'secondary is defined, secondaryLightRef may be needed. '
            'Set secondaryLightRef to the same color '
            'as secondary in your LIGHT theme mode setup to get correct values '
            'for "fixed" colors for your dark ColorScheme or to seed with '
            'same color as your light theme secondary. If secondary is already '
            'set to same color as light mode secondary and it is only used as '
            'seed in dark mode, defining secondaryLightRef is not necessary.');
      }
      if (tertiary != null &&
          tertiaryLightRef == null &&
          ((!seed.useKeyColors &&
                  (fixedColorStyle ?? FlexFixedColorStyle.computed) ==
                      FlexFixedColorStyle.computed) ||
              seed.useKeyColors)) {
        debugPrint('FlexColorScheme WARNING: tertiaryLightRef is null, but '
            'tertiary is defined, tertiaryLightRef may be needed. '
            'Set tertiaryLightRef to the same color '
            'as tertiary in your LIGHT theme mode setup to get correct values '
            'for "fixed" colors for your dark ColorScheme or to seed with '
            'same color as your light theme tertiary. If tertiary is already '
            'set to same color as light mode tertiary and it is only used as '
            'seed in dark mode, defining tertiaryLightRef is not necessary.');
      }
    }
```

If you have made your theme with **Themes Playground V8** you should typically **not** get into a situation where you see this information.

But if you make a totally custom theme using FlexColorScheme or just upgrade a Playground Theme from v7 to v8 that uses custom colors, then you may run into this warning.


In order for FlexColorScheme to be able to to compute or seed generate the colors for all the **ColorScheme** `Fixed` and `FixedDim` colors and their `on` colors in dark mode, so that they get the same color values as in light mode, which they according to M3 spec should have, a dark mode FlexColorScheme needs to know what colors where used as seed keys or input colors for computed/seeded version of these colors in the corresponding light theme.

The `primaryLightRef`, `secondaryLightRef` and `tertiaryLightRef` is the mechanism for providing this information to the dark mode theme. They should contain the color values for the used `primary`, `secondary` and `tertiary` colors in light mode. To be used either for seed generation of the fixed colors, or for their FCS computed version, depends on which version of fixed colors you have used.

When using built-in schemes, this is handled automatically, and when using Playground and making custom colored schemes with it, it also takes care of this for you.


![Screenshot 2024-12-13 at 18 18 05](https://github.com/user-attachments/assets/2b0fdbdd-6cc5-43f8-9789-7abd5912bfd3)

Color math and logics for M3 ColorScheme are bit complex, and got more confusing in Flutter 3.22 when so many new colors where added with specific roles.

If you are not using any of the fixed Colors in your app, meaning these:

![Screenshot 2024-12-13 at 18 21 43](https://github.com/user-attachments/assets/c960216a-e787-432e-b02c-7d6d4ce9a425)

Then you can safely ignore the informational warning.

By default Flutter's built-in components do not yet use these colors for anything either. They are just there to give you more options for your design by giving you a set of ColorScheme colors, that have the same color values in both light and dark theme mode.

See https://m3.material.io/styles/color/roles#26b6a882-064d-4668-b096-c51142477850

---

Your question and report was a bit short, so it was difficult to figure out what you wanted to know about it, or if you were just concerned that it was a bug.

In any case I hope this explains it and answers your questions.

I will convert this to a Question and Answer, as it it is likely a topic others will wonder about at some point too.

**Comment by maxfrees** [comment]
Thanks for the answer, each time your answer is very rewarding for us, and I'll try next week to see if I can add primaryLightRef to solve this prompt

**Comment by maxfrees** [comment]
@rydmike I have a problem, I want to remove the primaryLightRef warning message, I need to set colors, but the secondary in colors is mandatory, but I'm not sure what color the secondary should be set, my current primary Color is 0xFF0052D9,



![image](https://github.com/user-attachments/assets/069f31cd-8f82-4e01-93d2-b2a53d8f2e58)

**Reply by rydmike** [reply]
As tried to be shown in the screen shot above with the arrows, you set the dark theme mode color values as follows:

`primaryLightRef` = the color value of `primary` from light theme mode.
`secondaryLightRef` = the color value of `secondary` from light theme mode.
`tertiaryLightRef` = the color value of `tertiary` from light theme mode.

You don't have to do his (the warning is only shown in debug builds, not release). Why should you do this? For the light and dark theme mode ColorScheme to follow the M3 guide spec, where the fixed colors are equal in light and dark mode, we need to know the light mode colors in order for FlexColorScheme to be able to compute the same color values for the colors `primaryFixed`, `primaryFixedDim`, `onPrimaryFixed`, `onPrimaryFixedVariant`, `secondaryFixed`, `secondaryFixedDim`, `onSecondaryFixed`, `onSecondaryFixedVariant`, `tertiaryFixed`, `tertiaryFixedDim`, `onTertiaryFixed`, `onTertiaryFixedVariant` in dark mode as thy will get in light mode from the used `primary`, `secondary and `tertiary` colors. So it needs to know what colors you used for them in light mode.

the warning is there to let you know this. If you do not not specify the "Ref" colors in dark mode, it will use the `primary`, `secondary and `tertiary` from the dark mode to compute the fixed colors, these colors will be fine, but they won tbe the same color values as the colors the theme has in light mode. Their design intent is that they should be. If you d not use these colors it does not really matter. Flutter default component themes do not use them (yet at least).


To further demonstrate here is example, in code showing color value definitions and how they are assigned:

```dart
import 'package:flex_color_scheme/flex_color_scheme.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

/// The [AppTheme] defines light and dark themes for the app.
///
/// Theme setup for FlexColorScheme package v8.
/// Use same major flex_color_scheme package version. If you use a
/// lower minor version, some properties may not be supported.
/// In that case, remove them after copying this theme to your
/// app or upgrade package to version 8.1.0.
///
/// Use in [MaterialApp] like this:
///
/// MaterialApp(
///   theme: AppTheme.light,
///   darkTheme: AppTheme.dark,
///     :
/// );
abstract final class AppTheme {
  // Main custom light colors
  static const Color primaryLight = Color(0xFF4496E0);
  static const Color primaryContainerLight = Color(0xFFB4E6FF);
  static const Color secondaryLight = Color(0xFF202B6D);
  static const Color secondaryContainerLight = Color(0xFF99CCF9);
  static const Color tertiaryLight = Color(0xFF514239);
  static const Color tertiaryContainerLight = Color(0xFFBAA99D);
  //
  // The defined light theme.
  static ThemeData light = FlexThemeData.light(
    colors: const FlexSchemeColor(
      // Custom colors
      primary: primaryLight,
      primaryContainer: primaryContainerLight,
      secondary: secondaryLight,
      secondaryContainer: secondaryContainerLight,
      tertiary: tertiaryLight,
      tertiaryContainer: tertiaryContainerLight,
    ),
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
  // Main custom dark colors
  static const Color primaryDark = Color(0xFFB4E6FF);
  static const Color primaryContainerDark = Color(0xFF1E8FDB);
  static const Color secondaryDark = Color(0xFF99CCF9);
  static const Color secondaryContainerDark = Color(0xFF202B6D);
  static const Color tertiaryDark = Color(0xFFBAA99D);
  static const Color tertiaryContainerDark = Color(0xFF514239);

  // The defined dark theme.
  static ThemeData dark = FlexThemeData.dark(
    colors: const FlexSchemeColor(
      // Custom colors
      primary: primaryDark,
      primaryContainer: primaryContainerDark,
      primaryLightRef: primaryLight, // The light primary
      secondary: secondaryDark,
      secondaryContainer: secondaryContainerDark,
      secondaryLightRef: secondaryLight, // The light secondary
      tertiary: tertiaryDark,
      tertiaryContainer: tertiaryContainerDark,
      tertiaryLightRef: tertiaryLight, // The light tertiary
    ),
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      blendOnColors: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
}
```

This theme will look like this:

| Light theme | Dark theme |
| ---- | --- |
| ![Screenshot 2024-12-31 at 15 43 51](https://github.com/user-attachments/assets/239505b0-e2a0-4b66-82dc-fed16bb87826) | ![Screenshot 2024-12-31 at 15 43 59](https://github.com/user-attachments/assets/9a57a651-2ecb-4584-af28-cff864c12c43) | 

Where the fixed colors are equal in light and dark mode.

If you are using seed generated schemes, it assumes you are using the same seed color(s) for you light and dark generated `ColorScheme` and will not nag about the ref colors.


That's basically it. For some more guidance an ideas on how to quickly define you app's custom `ColorScheme` using just a few or even just one input color, read on below. 

**Warning**:  A lot of info! 😄 

## Using a lot less input colors?

Before **ColorScheme** seed generation was a thing a Material design and Flutter, FlexColorScheme already supported computational ColorScheme using very few input colors.

You can use the `FlexSchemeColor.from()` constructor to use much less input colors and get full `ColorScheme` based on just a single input light and dark primary color. This is same as using the `usedColors` properties in the Playground an dits API, but looks a bit more elegant since you do not even have to define the other colors.

Example:

```dart
abstract final class AppTheme {
  // Main custom light colors
  static const Color primaryLight = Color(0xFF4496E0);
  //
  // The defined light theme.
  static ThemeData light = FlexThemeData.light(
    colors: FlexSchemeColor.from(
      primary: primaryLight,
    ),
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
  // Main custom dark colors
  static const Color primaryDark = Color(0xFFB4E6FF);

  // The defined dark theme.
  static ThemeData dark = FlexThemeData.dark(
    colors: FlexSchemeColor.from(
      // Custom colors
      primary: primaryDark,
      primaryLightRef: primaryLight, // The light primary
    ),
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      blendOnColors: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
}
```

This theme will look like this:

| Light theme | Dark theme |
| ---- | --- |
| ![Screenshot 2024-12-31 at 15 56 24](https://github.com/user-attachments/assets/5a6b1356-3f82-4369-acca-9f065844565c) | ![Screenshot 2024-12-31 at 15 56 46](https://github.com/user-attachments/assets/5c29a577-f65d-43eb-8ca2-822a6561d60d) | 


### Just one input color and no seed generated ColorScheme?

FlexColorScheme even has a way f computing dark mode theme from the light mode input colors.

```dart
abstract final class AppTheme {
  // Main custom light colors
  static const Color primaryLight = Color(0xFF4496E0);
  //
  // The defined light theme.
  static ThemeData light = FlexThemeData.light(
    colors: FlexSchemeColor.from(
      primary: primaryLight,
    ),

    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    useMaterial3ErrorColors: true, // M3 error colors instead of M2
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
  // The defined dark theme.
  static ThemeData dark = FlexThemeData.dark(
    colors: FlexSchemeColor.from(
      primary: primaryLight,
    ).defaultError.toDark(45, false), // Compute all from primary light color
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      blendOnColors: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    useMaterial3ErrorColors: true, // M3 error colors instead of M2
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
}
```


The dark mode theme for this will look slightly different:

| Light theme | Dark theme |
| ---- | --- |
| ![Screenshot 2024-12-31 at 16 10 55](https://github.com/user-attachments/assets/e1294424-2d64-424f-b5dc-cfa29c0b79c6) | ![Screenshot 2024-12-31 at 16 11 14](https://github.com/user-attachments/assets/986f0001-726f-494d-8a3c-ff1fecf14c17) | 


The magic above it that all those shades where the result of single input color, call it the "brand" color on white paper, something companies often have, which works for the light mode, and all the other needed shades were computed, even the dark mode colors.

What you need to do of course all depend on the results you want.

### Use seed generation and fixed primary?

There are also ways to get ColorScheme from just one input color when using seed generation, that have even better shades and where the main primary "brand" color is used as primary in light mode and primaryContainer in dark mode. Usually the main brand color is saturated to work well on white paper. This config use that fact and re-uses the same brand on the appropriate colors in ligoht and dark mode, and computes very nice shades for everything else, that are still related to the main brand color.

Here is an example of that:

```dart
abstract final class AppTheme {
  // Main custom light color
  static const Color primaryLight = Color(0xFF4496E0);
  //
  // The defined light theme.
  static ThemeData light = FlexThemeData.light(
    colors: FlexSchemeColor.from(
      primary: primaryLight,
    ),
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    keyColors: const FlexKeyColors(
      keepPrimary: true,
    ),
    tones: FlexSchemeVariant.oneHue.tones(Brightness.light),
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
  // The defined dark theme.
  static ThemeData dark = FlexThemeData.dark(
    colors: FlexSchemeColor.from(
      primary: primaryLight,
      primaryContainer: primaryLight,
    ), // Compute all from primary light color
    subThemesData: const FlexSubThemesData(
      interactionEffects: true,
      tintedDisabledControls: true,
      blendOnColors: true,
      blendOnLevel: 32,
      useM2StyleDividerInM3: true,
      inputDecoratorIsFilled: true,
      inputDecoratorBorderType: FlexInputBorderType.outline,
      alignedDropdown: true,
      navigationRailUseIndicator: true,
      navigationRailLabelType: NavigationRailLabelType.all,
    ),
    keyColors: const FlexKeyColors(
      keepPrimaryContainer: true,
    ),
    tones: FlexSchemeVariant.oneHue.tones(Brightness.dark),
    visualDensity: FlexColorScheme.comfortablePlatformDensity,
    cupertinoOverrideTheme: const CupertinoThemeData(applyThemeToAll: true),
  );
}
```

The result of this is more balanced shades, while still keeping a defined main band color in focus and overall look matching it:



| Light theme | Dark theme |
| ---- | --- |
| ![Screenshot 2024-12-31 at 16 30 00](https://github.com/user-attachments/assets/4560d9da-5afd-4c2d-a64d-dc62c9c82356) | ![Screenshot 2024-12-31 at 16 30 12](https://github.com/user-attachments/assets/7e77b4cc-2401-4ddd-94ef-ad60a6f86294) |

**Comment by phdev-design** [comment]
my case add
const Color lightPrimaryColor = Color(0xFF1565C0); // Colors.blue[800]

**Reply by rydmike** [reply]
If that was you color for the primary color in light theme mode, and you want the dark mode "fixed" colors to follow the M3 spec and have the same color values, then yes that is the right choice.

**Reply by rydmike** [reply]
It is all explained pretty exhaustively above 😄
